### ISSUE TITLE
Stale job status after container reboot

### ISSUE THREAD

**Issue by mccalluc** [hubmapconsortium/user_workspaces_server#63]
On https://workspaces.api.hubmapconsortium.org

I'm not sure what the expected come is but I have a workspace with this status:
```
{
  "id": 11,
  "resource_job_id": 6259,
  "job_type": "JupyterLabJob",
  "status": "sleeping",
  "datetime_created": "2022-07-07T20:06:00.359Z",
  "datetime_start": null,
  "datetime_end": null,
  "core_hours": 0,
  "job_details": {
    "current_job_details": {
      "message": "Webserver ready.",
      "proxy_details": {
        "path": "/passthrough/4e8a2cd47a5c/11/lab?token=e55755...redacted...e8482",
        "port": 8888,
        "hostname": "localhost"
      },
      "connection_details": {
        "url_path": "/passthrough/4e8a2cd47a5c/11/lab?token=e55755...redacted...e8482",
        "url_domain": "https://workspaces-pt.hubmapconsortium.org"
      }
    },
    "request_job_details": {}
  }
}
```

When I try to access https://workspaces-pt.hubmapconsortium.org/passthrough/4e8a2cd47a5c/11/lab?token=e55755...redacted...e8482, the response is 500 with an empty body.

**Comment by jpuerto-psc** [comment]
@mccalluc TLDR: This is an issue with the job actually being dead, despite the API reporting otherwise.

What happened here is that job 11 was actually dead, not sleeping. Since the jobs (currently) run local to the docker container, when the docker container gets rebooted (or removed and re-built), the job dies. Also, since the job queue runs on redis, once the reboot occurs redis has no history of the jobs that were in the queue at the time of the reboot.

Few potential solutions here:

- Use a different queuing that has persistency across reboots (could just use postgres here since we already rely on it for DB operations).
- Any jobs that aren't dead (datetime_end is not NULL) can be loaded into the queue and re-checked to ensure that the job wasn't killed do to some power outage etc.

This issue "should" be avoidable once we switch over to using Slurm for job running (currently in progress). Will have to test to make sure that this is resolved by doing so.

**Comment by jpuerto-psc** [comment]
Update:

After making Redis persistent across reboots (using AOF and mounting directory from the VM on which docker is running) does not solve this issue. The issue seems to be as follows:

- DjangoQ (the queue that this project uses) kills any tasks that are pending in the queue when the docker container is restarted.
- Once DjangoQ starts back up, it treats those tasks as completely dead. IE, it does not attempt to put them back into the queue.
- This, in turn, means that those tasks will always report their status as whatever it was right before the reboot. Any "pending" tasks will stay as pending forever. This is the case even if the jobs are actually still running/pending, like with Slurm.

Potential solutions:
- Switch to a different queue broker?
- Not sure if feasible, but checking jobs on service start and throwing them into the queue if they are **not** in a complete/failed state. Issue with this might be if a job completed successfully while the service is down it would then report as "failed" if the job cannot be found...

**Comment by jpuerto-psc** [comment]
TLDR: Chose different broker which seemed to resolve the issue.

Retrospective:

There seemed to be a couple of issues at play here which, in conjunction, were causing this failure state. The largest part of this issue is that `docker-compose restart` does not seem to respect the `depends_on` entries in `docker-compose.yml`. This meant that when django-q was trying to kill its tasks and push them to the broker, it was unable to do so, ultimately leading to tasks getting completely lost.

In an attempt to resolve this, I explored the potential of queuing up jobs that were marked as either `running` or `pending` when starting up the Django server. However, this would only work when deploying this service as a container. The reason being that if tasks **aren't** lost (which is the situation when using `docker-compose restart`), then django-q would simply spawn duplicate task loops since those tasks were never actually lost.

I also explored trying to assign the task id for the currently executing task to the Job model. This would, in theory, allow me to check whether there is a task for that job in the queue; however, the task id is only available in the `hook` of the task chain, which would mean that it only is available once the task execution is complete. This would mean that there was no way to check whether a Job is **actively** being processed.

Another approach that was taken was trying to kill the cluster and its workers, purge the queue, restart the cluster and its workers, and then resubmit the jobs. However, starting the cluster from within the context of the Django app would not really work. That is, if someone where to deploy this using gunicorn with multiple (n) workers, there is a possibility that there would be multiple (n) clusters spawned as well.
 
Outside of exploring/implementing a completely different queuing system (not possible with time constraints), the best solution was choosing a different broker (the Django ORM) and making sure that `stop` and then `start` was used with docker-compose rather than `restart`. When doing so, jobs were updated after the containers were started back up.

This is something that has to be reviewed when more time is available, as there **will** be issues if the VM/machine on which the docker cluster is hosted on fails critically. Tasks would obviously be lost in this case, since there would be no time for DjangoQ to send its tasks to the broker.
